pkg/clusterconditions/promql/promql.go

```diff
@@ -0,0 +1,116 @@
+// Package promql implements a cluster condition based on PromQL queries.
+//
+// https://github.com/openshift/enhancements/blob/master/enhancements/update/targeted-update-edge-blocking.md#promql
+package promql
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/prometheus/client_golang/api"
+	prometheusv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"k8s.io/klog/v2"
+
+	"github.com/openshift/cluster-version-operator/pkg/clusterconditions"
+	"github.com/openshift/cluster-version-operator/pkg/clusterconditions/cache"
+)
+
+// PromQL implements a cluster condition that matches based on PromQL.
+type PromQL struct {
+	// Address holds the Prometheus query URI.
+	Address string
+
+	// HTTPClientConfig holds the client configuration for connecting to the Prometheus service.
+	HTTPClientConfig config.HTTPClientConfig
+}
+
+var promql = &cache.Cache{
+	Condition: &PromQL{
+		Address: "https://thanos-querier.openshift-monitoring.svc.cluster.local:9091",
+		HTTPClientConfig: config.HTTPClientConfig{
+			Authorization: &config.Authorization{
+				Type:            "Bearer",
+				CredentialsFile: "/var/run/secrets/kubernetes.io/serviceaccount/token",
+			},
+			TLSConfig: config.TLSConfig{
+				CAFile: "/etc/tls/service-ca/service-ca.crt",
+			},
+		},
+	},
+	MinBetweenMatches: 10 * time.Minute,
+	MinForCondition:   time.Hour,
+	Expiration:        24 * time.Hour,
+}
+
+// Valid returns an error if the condition contains any properties
+// besides 'type' and a valid `promql`.
+func (p *PromQL) Valid(ctx context.Context, condition *configv1.ClusterCondition) error {
+	if condition.PromQL == nil {
+		return errors.New("the 'promql' property is required for 'type: PromQL' conditions")
+	}
+
+	if condition.PromQL.PromQL == "" {
+		return errors.New("the 'promql.promql' query string must be non-empty for 'type: PromQL' conditions")
+	}
+
+	return nil
+}
+
+// Match returns true when the condition's PromQL evaluates to 1,
+// false when the PromQL evaluates to 0, and an error if the PromQL
+// returns no time series or returns a value besides 0 or 1.
+func (p *PromQL) Match(ctx context.Context, condition *configv1.ClusterCondition) (bool, error) {
+	clientConfig := api.Config{Address: p.Address}
+
+	if roundTripper, err := config.NewRoundTripperFromConfig(p.HTTPClientConfig, "cluster-conditions"); err == nil {
+		clientConfig.RoundTripper = roundTripper
+	} else {
+		return false, fmt.Errorf("creating PromQL round-tripper: %w", err)
```

Comment:
Error strings [should be lowercased][1].  And I think the fact that it is an error is already sufficient context, without having to actually say `Error` in the string.  For example, with [the caller's wrapping][2], this will render as:

```
Exposure to SomeRisk is unknown due to an evaluation failure: creating PromQL round-tripper: ...additional details...
...some message...
```

[1]: https://github.com/golang/go/wiki/CodeReviewComments#error-strings
[2]: https://github.com/openshift/cluster-version-operator/pull/663#discussion_r749646392

Reply:
In this case  having error in the message string makes it easier to understand.

Reply:
`Exposure to SomeRisk is unknown due to an evaluation failure: error creating PromQL round-tripper:`  looks better to me (with `error` string)